# Validation on Hierarchical Data

You are familiar with cross-validation and holdout-validation, but when should each of them be used? and what happens if our data has a complex structure?

In this notebook, we will demonstrate some examples to help answer these questions.

## Holdout validation

During holdout validation, we select a holdout sample at the beginning of the modelling process, before we have done any data processing or model training. This is how we test the performance of a model on a situation it has not seen before. If possible, it is good to use data which has also been collected separately, for example a separate year of data, or data from a seperate geographical sample. 

This helps ensure the relations in the model are *general* to the population and not *specific* to your dataset. 

It is vital that the validation data data remains untouched till testing!

Let's use the travel mode dataset to investigate scoring methods. In this dataset there isn't an easy way of selecting data which was collected sepearately, as we do not information such as the trip dates for instance. 

We will therefore have to rely on some sort of random sampling.

Let's load the data and the libraries we need.

In [1]:
import numpy as np
import pandas as pd

from sklearn.ensemble import RandomForestClassifier
from sklearn import metrics, preprocessing, model_selection

import matplotlib.pyplot as plt

%matplotlib inline

In [3]:
df = pd.read_csv('./data/travel_mode.csv')
bool_map = {'no': 0, 'yes': 1}
education_map = {'lower': 0, 'middle': 1, 'higher': 2}
income_map = {'less20': 0, '20to40': 1, 'more40': 2}
mode_map = {'car': 0, 'walk': 1, 'bike': 2, 'pt': 3}

df.mode_main = df.mode_main.map(mode_map)
df.male = df.male.map(bool_map)
df.license = df.license.map(bool_map)
df.weekend = df.weekend.map(bool_map)
df.education = df.education.map(education_map)
df.income = df.income.map(income_map)

df = pd.get_dummies(df)

#drop native column so n-1 catgories
df.drop('ethnicity_native', axis=1, inplace=True)

Let's have a look at the `household_id` column, along with some other columns.

In [4]:
df.head(20)

household_id  distance  density  age  male  education  income  cars  \
0           7875       1.0  1.26259   84     0          0       0     0   
1           7875      10.0  1.26259   84     0          0       0     0   
2          14449       3.0  1.76264   27     1          1       1     1   
3          14449       3.0  1.76264   27     1          1       1     1   
4          14449      61.5  1.76264   27     1          1       1     1   
5          14449      61.5  0.88025   27     1          1       1     1   
6          19452       0.5  1.83997   47     1          0       0     0   
7          19452       0.5  2.15367   47     1          0       0     0   
8          10039       0.4  1.18075   31     1          1       1     1   
9          10706       5.0  2.17375   65     0          0       0     1   
10         10706       5.0  1.25468   65     0          0       0     1   
11         10706       5.0  3.09575   65     0          0       0     1   
12         12465      20.0  0.70105   47     1          2       2     2   
13         12465      20.0  1.50813   47     1          2       2     2   
14          3460       1.0  2.61713   24     0          1       1     1   
15          3460       0.5  2.61713   24     0          1       1     1   
16          3460       0.5  5.26579   24     0          1       1     1   
17          3460       0.5  5.26579   24     0          1       1     1   
18         11644      25.2  2.79098   29     0          2       2     0   
19         11644      25.2  1.47212   29     0          2       2     0   

    license  bicycles  weekend  diversity      green  temp  precip  wind  \
0         1         1        1    1.24604  26.881233   0.1    0.10   3.0   
1         1         1        1    1.24604  26.881233   0.1    0.10   3.0   
2         1         2        1    1.53959  36.045955  -3.4    0.05   1.8   
3         1         2        1    1.53959  36.045955  -3.4    0.05   1.8   
4         1         2        1    1.53959  36.045955  -3.4    0.05   1.8   
5         1         2        1    1.53959  36.045955  -3.4    0.05   1.8   
6         1         3        1    1.72325  36.698977  -2.2    0.05   2.5   
7         1         3        1    1.72325  36.698977  -2.2    0.05   2.5   
8         1         2        1    1.33844  82.758365  -1.0    0.00   2.5   
9         1         2        1    1.36730  33.290612   0.2    2.20   1.8   
10        1         2        1    1.36730  33.290612   0.2    2.20   1.8   
11        1         2        1    1.36730  33.290612   0.2    2.20   1.8   
12        1         7        1    1.45511  85.867899  -0.8    0.05   2.7   
13        1         7        1    1.45511  85.867899  -0.8    0.05   2.7   
14        1         3        1    1.02295  22.163588  -1.0    0.00   2.5   
15        1         3        1    1.02295  22.163588  -1.0    0.00   2.5   
16        1         3        1    1.02295  22.163588  -1.0    0.00   2.5   
17        1         3        1    1.02295  22.163588  -1.0    0.00   2.5   
18        0         3        1    1.41009  38.285714  -0.5    0.00   2.8   
19        0         3        1    1.41009  38.285714  -0.5    0.00   2.8   

    mode_main  ethnicity_nonwestern  ethnicity_western  
0           1                 False              False  
1           1                 False              False  
2           0                 False               True  
3           0                 False               True  
4           0                 False               True  
5           0                 False               True  
6           1                 False              False  
7           1                 False              False  
8           1                 False              False  
9           0                 False              False  
10          0                 False              False  
11          0                 False              False  
12          0                 False              False  
13          0                 Fal

So we can clearly see that people from the same household seem to be making multiple trips, and they will be highly correlated, in terms of explanitary variables (e.g. green, the percentage of green space in the vicinity around the household), and the target variable (e.g. mode choice). 

For instance, four trips are made a 24 year old male in household 3460, all walking, and all with identical explanitary variables, except distance and density.

This data is therefore *hierarchical*.

### Implications of hierarchical data

So what is the relevance of the hierarchical nature of the data?

When we sample data randomly to form a test set, each individual row has an equal probability of being selected, which means there is a high chance of rows (trips) made by the same household appearing in both the test and train dataset. This is bad, as we can then overfit the model to noise in that households data - i.e. unique features of the household/individual, and not general relations which indicate someone is likely to take one mode over another.

Let's dig a little deeper to find out more.

In [5]:
print('There are {} trips made by {} households in the data'.format(len(df), max(df.household_id)))

There are 230608 trips made by 45036 households in the data


So it looks like we have around 5 trips made by each household! For sure that will introduce correlations between test and train data which are specific to the household, and not general.

First, let's generate some results for a model that we know will overfit, a decision tree with no maximum depth, using our original random sampling. First lets generate train and test folds. We will add the suffix `_r` for random.

In [6]:
y = df.mode_main
hh = df.household_id
X = df.drop(['mode_main', 'household_id'], axis=1)

In [7]:
X_train_r, X_test_r, y_train_r, y_test_r = model_selection.train_test_split(X, y)

Then we can create our candidate model. We'll use standard parameters, except we wont restrict the tree depth!

In [8]:
clf_r = RandomForestClassifier(n_estimators=50, max_features=3, max_depth=None, n_jobs=-1)

The next cell might be slow!

In [9]:
clf_r.fit(X_train_r, y_train_r)

RandomForestClassifier(max_features=3, n_estimators=50, n_jobs=-1)

Next, caluclate the discrete and probabilistic classifications for the classifier, and use them to calculate the predicted log loss and accuracy.

In [10]:
# generate y_pred and y_probs
y_pred_r = clf_r.predict(X_test_r)
y_probs_r = clf_r.predict_proba(X_test_r)


In [11]:
# and the accuracy score
metrics.accuracy_score(y_test_r, y_pred_r)

0.8258169707902588

In [13]:
# and the log loss
metrics.log_loss(y_test_r, y_probs_r)


0.6127672653174836

Great, we have some baseline scores to compare against! By not restricing the tree depth at all, we are almost definitely overfitting our model to the noise from the hierarchical data. We can see our metrics have improved hugely from yesterday, just by not restricting the tree depth. 

We are essenitally predicting the training data!

### Grouped sampling
But how do we deal with hierarchical data when sampling test data?

Scikit learn can help! We can use `GroupShuffleSplit` from the `model_selection` module, see the documentation [here](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GroupShuffleSplit.html).

Remember, we only need one split for holdout-validation, so be sure to set the approriate parameters. Also stick to a 70:30 split.

Also, when using split, make sure to use the household series (stored as `hh`) as the groups.

*HINT* `split` returns a generator, so you will need to get the result out of the generator by iterating over it. Remember, you can use [`next`](https://docs.python.org/2/library/functions.html#next) to simplify things!

In [16]:
from sklearn.model_selection import GroupShuffleSplit
gss = GroupShuffleSplit(n_splits=1, train_size= 0.7, random_state= 42)

train_idx, test_idx = next(gss.split(X, y, groups=hh))

We can then use the indices from the split to generate the train and test data (yes it is a little longwinded!) We can use the suffix `_g` for grouped.

In [17]:
X_train_g, X_test_g, y_train_g, y_test_g = X.iloc[train_idx], X.iloc[test_idx], y[train_idx], y[test_idx]

Now lets fit the same classifier to our correctly sampled data...

In [18]:
clf_g = RandomForestClassifier(n_estimators=50, max_features=3, max_depth=None, n_jobs=-1)
clf_g.fit(X_train_g, y_train_g)

RandomForestClassifier(max_features=3, n_estimators=50, n_jobs=-1)

In [19]:
y_pred_g = clf_g.predict(X_test_g)
y_probs_g = clf_g.predict_proba(X_test_g)

In [20]:
metrics.accuracy_score(y_test_g, y_pred_g)

0.6777958936326778

In [21]:
metrics.log_loss(y_test_g, y_probs_g)

0.9007184775169212

Wow! That's a pretty big drop in our metrics - accuracy has dropped to from 83% to 68%! Goes to show it is **very** important to deal with hierarchical data properly.

### Cross-validation

Cross validation is when we can estimate model performance on a dataset of fixed size. It is very useful for instance when trying to select model parameters.

It is important to note, that the validation we do to select model parameters/feature engineering, needs to be completely sepearate to the holdout-validation to test the model. For instance, if we separate data into a testing and a holdout training set, we should then only do model selction using the training set. Otherwise, we will be selecting parameters which allow us to fit to our testing data. This is known as *data leakage*.

So how do we do cross-validation with our hiearchical data? This time, we can use [GroupKFold](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GroupKFold.html )!

We will use GroupKFold to do a simple parameter search on max-depth for the forest.

Firstly, we need to get hold of the *groups* (`household_id`) for the train data.

In [ ]:
gss = model_selection.GroupShuffleSplit(n_splits=1, train_size=.7, test_size=.3, 
                                        random_state=42)
train_idx, test_idx = next(gss.split(X, y, groups=hh))
X_train_g, X_test_g, y_train_g, y_test_g = X.iloc[train_idx], X.iloc[test_idx], y[train_idx], y[test_idx]
hh_train_g, hh_test_g = hh.iloc[train_idx], hh.iloc[test_idx]

Next we can use `GroupKFold` with `GridSearchCV`, documented [here](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html).

*HINT* Use the `cv` parameter within `GridSearchCV` to do a grouped K-fold CV.

Let's leave the other parameters the same, and just investigate the `max_depth` parameter. We can try values of 2,5 and 10, as a crude search.

In [ ]:
clf_cv = RandomForestClassifier(n_estimators=50, max_features=3, n_jobs=-1)

In [ ]:
params = {'max_depth': [2, 5, 10]}
# use GroupKFold with 3 splits and GridSearchCV to search max_depth valus of 2, 6 and 12.



So it seems the `max_depth` of 12 worked best out of the values we tried. Let's see how well this classifier works on our validation data.

In [ ]:
clf_cv = gs.best_estimator_

In [ ]:
clf_cv_pred = gs.predict(X_test_g)
clf_cv_probs = gs.predict_proba(X_test_g)

In [ ]:
metrics.accuracy_score(y_test_g, clf_cv_pred)

In [ ]:
metrics.log_loss(y_test_g, clf_cv_probs)

Finally, we can investigate which features were most important in the two different classifiers. The following cells plot bar charts of the feature importances for each classifier.

In [ ]:
features = X.columns
importances = clf_r.feature_importances_
indices = np.argsort(importances)

plt.title('Feature Importances')
plt.barh(range(len(indices)), importances[indices], color='b', align='center')
plt.yticks(range(len(indices)), features[indices])
plt.xlabel('Relative Importance')
plt.show()

In [ ]:
features = X.columns
importances = clf_cv.feature_importances_
indices = np.argsort(importances)

plt.title('Feature Importances')
plt.barh(range(len(indices)), importances[indices], color='b', align='center')
plt.yticks(range(len(indices)), features[indices])
plt.xlabel('Relative Importance')
plt.show()

So it seems the first classifier, with the random sampling, was overfitting to the `float` values, `density`, `green`, and `diversity`. This makes sense, as each tree could just memorise the mode associated with each unique value, and repeat it for the test data. The classifier using grouped sampling puts much more emphasis on car ownership and trip distance.